### DISCUSSION TITLE
Container host HA and floating IPs

### DISCUSSION THREAD

**Discussion by Adam7288** [BretFisher/ama#210] [General]
Hi Bret,

Thanks for your awesome course on docker / k8s. 

One thing I cannot wrap my head around is how to achieve HA if the route to a static ip fails. Let's say I have one of the nodes of a k8s 3-node cluster set to a static ip address and in my public nameserver I have all 3 ips listed for domain abc.com.

I've heard of very low TTL values, or DNS round robin (that is discouraged because of local caching of dns results) but how is it actually done in production for zero downtime?

I've googled like crazy and best I can gather is some kind of "floating ip" solution. But that still does not provide an explanation to me. There is some fundamental lack of understanding on my part.

You don't have to give me a full blown write up, maybe just some key terms to conduct further research on would be super appreciated.

**Comment by BretFisher** [comment]
You're right, floating IP solutions do exist in traditional Linux OSs (see [heartbeat and pacemaker](https://en.wikipedia.org/wiki/Linux-HA)), but at this point they are legacy in the cloud and kubernetes. They are just too complex to deal with usually when you don't have deep control over your networking hardware. Typical options today:

1. cloud load balancers in front of your cluster. This is how all the clouds do it, and the most likely solution for people using K8s in aws/azure/google etc. In Kubernetes resources, this would be a Service of type LoadBalancer that is really K8s controlling a cloud LB external to itself. They route traffic typically to a Service of type NodePort (on any K8s node) or if you're lucky, directly to the Service ClusterIP, which will save a tiny step in the connection process.
2. If in a datacenter, you'll likely need a dedicated hardware LB (and they come with their own redundancy) like a [BIG-IP from F5](https://www.f5.com/solutions/load-balancing-your-applications). This sits in front of clusters and control/route all incoming client traffic to the proper Service NodePorts that are listening on all nodes.
3. If you don't have an F5, and can control your networking hardware, then [MetalLB](https://metallb.universe.tf/) may be an option. It's kind of the Kubernetes-friendly way to do heartbeat/pacemaker. The [concepts page](https://metallb.universe.tf/concepts/) describes how it works quite well, assuming you know your IP and routing concepts well.
